userapi/src/main/java/fk/prof/userapi/api/ProfileStoreAPIImpl.java

```diff
@@ -150,88 +148,80 @@ public void getProfilesInTimeWindow(Future<List<AggregatedProfileNamingStrategy>
+            vertx.setTimer(loadTimeout, timerId -> timeoutRequestedFutureInMap(fileNameKey, future, filesBeingLoaded));
 
-            if (!fileLoadInProgress) {
+            if(!fileLoadInProgress) {
                 workerExecutor.executeBlocking((Future<AggregationWindowSummary> f) -> profileLoader.loadSummary(f, filename),
-                        true,
-                        result -> completeAggregatedProfileLoading(summaryCache, result, fileNameKey));
+                    true,
+                    result -> {
+                        if(result.succeeded()) {
+                            summaryCache.put(fileNameKey, result.result());
+                        }
+                        filesBeingLoaded.remove(fileNameKey).complete(result.map(r -> r));
+                    });
             }
         } else {
             future.complete(cachedProfileInfo);
         }
     }
 
+    @Override
+    public <T extends ProfileView> Future<Pair<AggregatedSamplesPerTraceCtx, T>> getProfileView(AggregatedProfileNamingStrategy profileName, String traceName, ProfileViewType profileViewType) {
+        Future<Pair<AggregatedSamplesPerTraceCtx, T>> result = Future.future();
+
+        Future<Pair<AggregatedSamplesPerTraceCtx, T>> f = clusterAwareCache.getProfileView(profileName, traceName, profileViewType);
+        f.setHandler(ar -> {
+            if (ar.failed() && ar.cause() instanceof CachedProfileNotFoundException && !((CachedProfileNotFoundException) ar.cause()).isCachedRemotely()) {
+                // initiate the load locally
+                Future<AggregatedProfileInfo> profileLoad = Future.future();
+                load(profileLoad, profileName);
+
+                profileLoad.setHandler(ar2 -> {
+                    Future<Pair<AggregatedSamplesPerTraceCtx, T>> f2 = clusterAwareCache.getProfileView(profileName, traceName, profileViewType);
+                    f2.setHandler(result.completer());
+                });
+            } else {
+                result.handle(ar);
+            }
+        });
+        return result;
+    }
 
-    synchronized private <T> void saveRequestedFuture(String filename, Future<T> future) {
-        FuturesList futures = futuresForLoadingFiles.get(filename);
+    private <T> void saveRequestedFutureInMap(String filename, Future<T> future, Map<String, FuturesList<T>> futuresListMap) {
+        FuturesList<T> futures = futuresListMap.get(filename);
         if (futures == null) {
-            futures = new FuturesList();
-            futuresForLoadingFiles.put(filename, futures);
+            futures = new FuturesList<>();
+            futuresListMap.put(filename, futures);
         }
         futures.addFuture(future);
     }
 
-    synchronized private <T> void timeoutRequestedFuture(String filename, Future<T> future) {
+    private <T> void timeoutRequestedFutureInMap(String filename, Future<T> future, Map<String, FuturesList<T>> filesBeingLoaded) {
```

Comment:
Shouldn't a method like this be member of FuturesList -> `timeoutFuture(Future<T> future, Throwable th)`?
Also that method should be synchronized like other methods in FuturesList. This can avoid the race condition that exists here when a future is failed because of timeout and concurrently `FuturesList.complete()` is called

Reply:
A good way to think is that complete/fail operations on futures belonging to FuturesList should synchronize on the same object to avoid races

Reply:
Here, FutureList is nothing but a synchornized list. I dont need the atomicity on each element of this list. 
And I didn't  used the synchronized list itself because i needed atomic addIfAbsent.